### ISSUE TITLE
Auto bootstrapping after outage does not work

### ISSUE THREAD

**Issue by fidian** [hashicorp/consul#856]
This is likely related to #526.

I have three Consul servers and all are talking with Atlas.  They are automatically bootstrapped and it's set to a quorum of 3.  I proceed to kill two of them.  I happened to kill the cluster leader.  Now I get messages like this:

    2015/04/10 15:23:49 [ERR] agent: failed to sync remote state: No cluster leader
    2015/04/10 15:23:58 [WARN] dns: Query results too stale, re-requesting
    2015/04/10 15:23:58 [ERR] dns: rpc error: No cluster leader
    2015/04/10 15:23:58 [WARN] dns: Query results too stale, re-requesting
    2015/04/10 15:23:58 [ERR] dns: rpc error: No cluster leader
    2015/04/10 15:23:58 [WARN] dns: Query results too stale, re-requesting
    2015/04/10 15:23:58 [ERR] dns: rpc error: No cluster leader
    2015/04/10 15:23:58 [WARN] dns: Query results too stale, re-requesting
    2015/04/10 15:23:58 [ERR] dns: rpc error: No cluster leader
    2015/04/10 15:23:58 [WARN] dns: Query results too stale, re-requesting
    2015/04/10 15:23:58 [ERR] dns: rpc error: No cluster leader

Ok, I figure that perhaps there's some sort of split-brain code thinking there's a problem.  I spin up two new servers (all are made from the same AMI in Amazon).  Now I have three servers all complaining.

*Issue 1:* Isn't the use of Atlas supposed to fix this problem?  Wouldn't the lonely survivor talk to Atlas and realize that it should be its own cluster leader?  Shouldn't a connection to Atlas be deemed important enough and let Atlas decide who is the new cluster leader?  The old one has no connection any longer and whomever is talking to Atlas should possibly take control?

*Issue 2:* When I spin up two more servers, shouldn't the group realize they're the only ones and then they should automatically bootstrap the cluster?

I guess I'm really confused with how this automatic bootstrapping is supposed to work and how Atlas helps to solve the problem.

**Comment by armon** [comment]
This sounds like the cluster has entered into an outage scenario (quorum lost). There are some steps on solving this here: https://consul.io/docs/guides/outage.html

To answer the other question about Atlas, there are really a few different bootstrapping problems at play. The problem Atlas tries to solve is how does an agent discover the address of a peer in the cluster. Especially in elastic environments where clients are spinning up and down automatically, Atlas solves that problem by triggering a "consul join" automatically.

Then there is the problems around establishing and maintaining a quorum. Using `-bootstrap-expect` along with `-atlas-join` solves the establishing problem. Maintaining that quorum is done by Consul automatically by allowing servers to join/leave while quorum is possible.

However, the moment quorum is lost and there is existing data in the cluster, there is the missing gap of re-establishing quorum. This is a much much trickier problem from a safety standpoint. We've discussed trying to handle this via Atlas, and it is certainly do-able, but ultimately there are a number of risky situations that could cause data loss preventing us from doing it currently. Safely recovering from an outage requires a lot of context which an operator is likely to have, but not necessarily information we have access to (are the servers actually alive, is it a temporary problem, firewall settings, network settings, etc).

This is why we currently leave that process as manual and document the recovery steps. In standard operation of Consul, this should be an exceedingly rare event. It requires either an operational mistake or multiple server failures before an operator is able to boot new servers to replace the failed ones.

So TL;DR: it is a problem and a missing gap we are aware of, but there are fundamental safety issues to address before we would be comfortable tackling this in an automatic way.

**Comment by fidian** [comment]
I can understand the difficulty - thanks for the in-depth explanation!
